### ISSUE TITLE
Make diff_file resilient to line-ending differences on Windows

### ISSUE THREAD

**Issue by tannervoas742** [mesonbuild/meson#12092]
**Question**

I have systems that have patch.exe plus git and others which only have git. I additionally have a diff_file which successfully is applied when only git is present but fails on systems which have patch.exe. It appears wrap files are defaulting to using patch.exe if its present but I want the opposite behavior; always use git. How do I tell a wrap file to always apply the diff_file using git?

**Comment by tannervoas742** [comment]
> There is currently no way to do so. Can you provide an example of a patch that fails with patch but works with git apply ?

Its a very simple patch. Not sure why its failing to apply through patch.exe. I do not have experience with patch.exe though. Does it expect a different format than git?

```cpp
--- a/inc/loader/example.h
+++ b/inc/loader/example.h
@@ -362,7 +362,7 @@ protected:
 
         wcscpy_s(dllFullName, bufferLength, dllPath);
         wcscat_s(dllFullName, bufferLength, libraryName);
-        file_handle = ::LoadLibrary(dllFullName);   //try to load 32-bit lib
+        file_handle = ::LoadLibraryW(dllFullName);   //try to load 32-bit lib
 
 #endif
     }
@@ -442,4 +442,4 @@ protected:
     examplePfn exampleGetVersion;
 };
 
-#endif //_EXAMPLE_H_
\ No newline at end of file
+#endif //_EXAMPLE_H_

```

**Comment by kiwixz** [comment]
Looks pretty straightforward to me and worked with my `patch -p1`...  Did you make sure the line endings are the same ?

**Comment by xclaesse** [comment]
What error do you get? I don't think there is a downside in using git by default. In the case the patch is being applied on a git repository, I think it has more advanced strategies for resolving conflicts automatically.

**Comment by xclaesse** [comment]
Interestingly, I just hit this issue when building GStreamer on Windows. [GLib patches](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/tree/main/subprojects/packagefiles/glib-2.76.4) they have did not apply with  Perl's patch.exe!

@eli-schwartz this could be yet another instance of Perl messing with people's system...

**Comment by eli-schwartz** [comment]
Can someone delete Strawberry Perl from the internet?

**Comment by tannervoas742** [comment]
I confirmed its the line endings. Changing the patch file to CRLF fixes patch.exe case, but breaks git case. Anyways this is now an issue that I can resolve internally in my project. This ticket can be closed.

**Comment by eli-schwartz** [comment]
Well, I think there's a decent chance we need to hard mask perl's version of patch.exe, if it's going to behave unusually surprisingly.

**Comment by dvrogozh** [comment]
> Well, I think there's a decent chance we need to hard mask perl's version of patch.exe

Not only perl's version of patch.exe. Some other software may include patch.exe which breaks things. For example, Beyond Compare is coming with such version of patch.exe.

Problem is to make patching of subprojects working across different environment people might potentially have. And accounting for both patch and git environments at the same time gives issues. Which is I can make perl's patch work by making sure that .patch has crlf, but this breaks patching with git. And vice versa - if I manage git patching work we have this story with perl's patch.

Does this make sense to extend wrap syntax to be able to clearly state how patching should be done?
```
[wrap-file]
<...>
diff_files = titansdk/0001-Fix-qcs-library-loading-at-32-bit-Windows.patch
diff_apply_with = patch  # or git
```

**Comment by dvrogozh** [comment]
Ok, so on my side I handled this problem by stepping out from `diff_files`. Instead of applying patches, I just rely on overlay `patch_directory` and replace file(s) which need patching with actual version I want to get.

Eventually that's not an optimal solution and it would be very nice to be able to use patches. However, as of today this support highly depends on user's environment which hard to predict. I think this needs to be improved to allow better meson scaling on Windows.

**Comment by tristan957** [comment]
@tannervoas742 can you post the help text for the Strawberry Perl patch.exe, so we can mask it?

**Comment by tristan957** [comment]
> For example, Beyond Compare is coming with such version of patch.exe.

@dvrogozh post the help output here too.

**Comment by dvrogozh** [comment]
>> For example, Beyond Compare is coming with such version of patch.exe.
>
> @dvrogozh post the help output here too, unless it is the same.

FYI, I was using Beyond Compare Version 4.4 (build 25886).

patch version:
```
$ "/c/Program Files/Beyond Compare 4/Patch.exe" --version
patch 2.5.9
Copyright (C) 1988 Larry Wall
Copyright (C) 2003 Free Software Foundation, Inc.

This program comes with NO WARRANTY, to the extent permitted by law.
You may redistribute copies of this program
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.

written by Larry Wall and Paul Eggert
```
And help:
```
$ "/c/Program Files/Beyond Compare 4/Patch.exe" --help
Usage: C:\Program Files\Beyond Compare 4\Patch.exe [OPTION]... [ORIGFILE [PATCHFILE]]

Input options:

  -p NUM  --strip=NUM  Strip NUM leading components from file names.
  -F LINES  --fuzz LINES  Set the fuzz factor to LINES for inexact matching.
  -l  --ignore-whitespace  Ignore white space changes between patch and input.

  -c  --context  Interpret the patch as a context difference.
  -e  --ed  Interpret the patch as an ed script.
  -n  --normal  Interpret the patch as a normal difference.
  -u  --unified  Interpret the patch as a unified difference.

  -N  --forward  Ignore patches that appear to be reversed or already applied.
  -R  --reverse  Assume patches were created with old and new files swapped.

  -i PATCHFILE  --input=PATCHFILE  Read patch from PATCHFILE instead of stdin.

Output options:

  -o FILE  --output=FILE  Output patched files to FILE.
  -r FILE  --reject-file=FILE  Output rejects to FILE.

  -D NAME  --ifdef=NAME  Make merged if-then-else output using NAME.
  -E  --remove-empty-files  Remove output files that are empty after patching.

  -Z  --set-utc  Set times of patched files, assuming diff uses UTC (GMT).
  -T  --set-time  Likewise, assuming local time.

  --quoting-style=WORD   output file names using quoting style WORD.
    Valid WORDs are: literal, shell, shell-always, c, escape.
    Default is taken from QUOTING_STYLE env variable, or 'shell' if unset.

Backup and version control options:

  -b  --backup  Back up the original contents of each file.
  --backup-if-mismatch  Back up if the patch does not match exactly.
  --no-backup-if-mismatch  Back up mismatches only if otherwise requested.

  -V STYLE  --version-control=STYLE  Use STYLE version control.
        STYLE is either 'simple', 'numbered', or 'existing'.
  -B PREFIX  --prefix=PREFIX  Prepend PREFIX to backup file names.
  -Y PREFIX  --basename-prefix=PREFIX  Prepend PREFIX to backup file basenames.
  -z SUFFIX  --suffix=SUFFIX  Append SUFFIX to backup file names.

  -g NUM  --get=NUM  Get files from RCS etc. if positive; ask if negative.

Miscellaneous options:

  -t  --batch  Ask no questions; skip bad-Prereq patches; assume reversed.
  -f  --force  Like -t, but ignore bad-Prereq patches, and assume unreversed.
  -s  --quiet  --silent  Work silently unless an error occurs.
  --verbose  Output extra information about the work being done.
  --dry-run  Do not actually change any files; just print what would happen.
  --posix  Conform to the POSIX standard.
  --unified-reject-files  Create unified reject files.
  --global-reject-file=file  Put all rejects into one file.

  -d DIR  --directory=DIR  Change the working directory to DIR first.
  --binary  Read and write data in binary mode.

  -v  --version  Output version info.
  --help  Output this help.

Report bugs to <bug-patch@gnu.org>.
```

**Comment by tristan957** [comment]
Doesn't seem to be anything particularly unique there like what we do in https://github.com/mesonbuild/meson/blob/d2dfef5205f031ab10591899a1fc5048ea6f4134/mesonbuild/dependencies/pkgconfig.py#L216

**Comment by dvrogozh** [comment]
> @tannervoas742 can you post the help text for the Strawberry Perl patch.exe, so we can mask it?

Tanner will comment later tonight, but I think he stepped into the issue using Beyond Compare, same as me. First comment regarding Strawberry Perl came from @xclaesse, see https://github.com/mesonbuild/meson/issues/12092#issuecomment-1669749503.

**Comment by dvrogozh** [comment]
> Doesn't seem to be anything particularly unique there like what we do in

Unfortunately yes. It looks to me as ordinary patch which I have in MSYS, just older version. Could there be some bug in older patch?

**Comment by tristan957** [comment]
I have no clue. If you pull the git repo for patch, I guess you could do a bisect and find the version that passes.

**Comment by dvrogozh** [comment]
> If you pull the git repo for patch, I guess you could do a bisect and find the version that passes.

Yeah, probably someone will need to do that to be sure. (I hope that BC did not use some weird build options and/or custom patches...)

**Comment by dvrogozh** [comment]
Ok, so I tried to build patch from sources (see https://savannah.gnu.org/git/?group=patch). And it did NOT work to apply a patch. While MSYS2 of the same version (2.7.6) does work just fine. Thus, I actually think that these are custom MSYS2 patches which make GNU patch functional in our use case. In particular, I believe at least this line is responsible for the magic:

https://github.com/msys2/MSYS2-packages/blob/7acec1f597e9b9df8585501db7b03f4c18782ab7/patch/msys2-patch-2.7.1.patch#L163

Note: I built patch from sources under MSYS2 MSYS environment with:
```
./bootstrap
./configure
make
```

**Comment by xclaesse** [comment]
IMHO the easiest way to get at least consistent behavior cross platforms is to rely on git by default: https://github.com/mesonbuild/meson/pull/12248.

**Comment by dvrogozh** [comment]
> IMHO the easiest way to get at least consistent behavior cross platforms is to rely on git by default

Fully agree.

**Comment by nirbheek** [comment]
> What error do you get? I don't think there is a downside in using git by default. In the case the patch is being applied on a git repository, I think it has more advanced strategies for resolving conflicts automatically.

This actually broke a project I'm working on, in which the diff_file applies to a submodule of the git subproject being cloned ([sentry](https://github.com/getsentry/sentry-native.git), in this case). `git` cannot correctly apply the patch in that case, but `patch` can.

I think we should revert this patch, and fix the original line-endings issue with `patch --ignore-whitespace` or something.

**Comment by antonysigma** [comment]
> Can someone delete Strawberry Perl from the internet?

Are you referring to this commit on Wrapdb CI jobs? https://github.com/fennewald/wrapdb/commit/3ba1b921c8f7a33caf8a93af2a050b5dca47127e

**Comment by Totto16** [comment]
Exactly the same problem here, with patch (strawberry perl) as default it worked, with git as default it doesn't. The file has LF line endings, since it's used cross platform, it is sdl2, with some custom patches, that are applied, and only failed after git was default, it runs on Github CI. It's really annoying, since I have to patch the CI every time 😓 I didn't investigate the   exact error git has with the patch 😓

**Comment by mikedickey** [comment]
Sorry for the github reference spam; I just noticed my force-pushes while trying to get this to work were adding comments here.

+1 for reverting the patch to prefer `git` instead. I think it's made things worse. `git` is just as flaky, at least the version provided by `choco` which is used by GitHub's CI/CD Windows runners has all the same problems discussed here. I couldn't figure out a clean way to "fix" `git` but was able to "fix" `patch` on them, so I opted to pin meson to an older version that still prefers `patch` instead.
